packages/polaris-viz/src/components/TooltipWrapper/hooks/useRepositionTooltip.ts

```diff
@@ -0,0 +1,337 @@
+      tooltipDimensions,
+      x,
+      y,
+    }: {
+      chartType: InternalChartType;
+      seriesBounds: BoundingRect | null;
+      tooltipDimensions: Dimensions | null;
+      x: number;
+      y: number;
+    }) => {
+      if (
+        tooltipDimensions == null ||
+        tooltipDimensions.width === 0 ||
+        tooltipDimensions.height === 0
+      ) {
+        return {x, y};
+      }
+
+      let newX = x;
+      let newY = y;
+
+      if (isPerformanceImpacted) {
+        newY = containerBounds.y;
+      }
+
+      if (isTouchDevice) {
+        newY = containerBounds.y - tooltipDimensions.height;
+      }
+
+      // Reposition the tooltip for different chart types.
+      // Example: BarChart renders the tooltip centered and above
+      // the data series.
+      switch (chartType) {
+        case InternalChartType.Bar:
+          newY -= tooltipDimensions.height + TOOLTIP_SPACING;
+          newX -= tooltipDimensions.width / 2;
+          break;
+        case InternalChartType.HorizontalBar:
+          newX += TOOLTIP_SPACING;
+          newY -= tooltipDimensions.height / 2;
+          break;
+        default:
+          newY -= tooltipDimensions.height / 2;
+          newX += TOOLTIP_SPACING;
+          break;
+      }
+
+      // Once the tooltip is positioned, we need to ensure that
+      // it's not rendering outside the browser window,
+      // or covering any data series.
+      return findOptimalPosition({
+        chartType,
+        scrollContainer,
+        seriesBounds,
+        tooltipDimensions,
+        x: newX,
+        y: newY,
+      });
+    },
+    [containerBounds, isPerformanceImpacted, scrollContainer, isTouchDevice],
+  );
+}
+
+function findOptimalPosition({
+  chartType,
+  scrollContainer,
+  seriesBounds,
+  tooltipDimensions,
+  x,
+  y,
+}: {
+  chartType: InternalChartType;
+  scrollContainer?: Element | null;
+  seriesBounds: BoundingRect | null;
+  tooltipDimensions: Dimensions;
+  x: number;
+  y: number;
+}): Position {
+  if (
+    isPositionWithinWindowAndOutsideSeriesBounds({
+      scrollContainer,
+      seriesBounds,
+      tooltipDimensions,
+      x,
+      y,
+    })
+  ) {
+    return {x, y};
+  }
+
+  // Check if position would be outside window bounds
+  const tooltipBounds = {
+    top: y,
+    left: x,
+    bottom: y + tooltipDimensions.height,
+    right: x + tooltipDimensions.width,
+  };
+
+  const isOutsideWindow = !isWithinWindowBounds({
+    ...tooltipBounds,
+    scrollContainer,
+  });
+
+  const {containerWidth, containerHeight} =
+    getScrollContainerBounds(scrollContainer);
+
+  const containerMax =
+    containerWidth - tooltipDimensions.width - SCROLLBAR_SIZE;
+  let newX = x;
+
+  // If the tooltip has been repositioned where it would render
+  // outside the browser window, or cover the data series,
+  // we want to reposition it again around the series bounds.
+  if (isOutsideWindow && seriesBounds != null) {
+    const positions: Position[] = [
```

Comment:
For charts that don't update their position, we likely pass a `seriesBounds` prop.

This prop is a set of `x, y, height and width` values that we use to position the tooltip around.

For example, in a BarChart, if the tooltip has been repositioned into the `seriesBounds` we will try and reposition the tooltip to the left of the `seriesBounds` so that the tooltip does not cover the bars.